### ISSUE TITLE
SWA CLI stopping without error msg when using Blazor WASM dev server over https

### ISSUE THREAD

**Issue by StacyCash** [Azure/static-web-apps-cli#610]
**Are you accessing the CLI from the default port `:4280` ?**

- [ ] No, I am using a different port number (`--port`) and accessing the CLI from that port
- [X] Yes, I am accessing the CLI from port `:4280`

**Describe the bug**
When trying to use the SWA CLI with a dotnet Blazor application running on https the SWA CLI quits when the index page is accessed (see below for traces from various node versions)

When accessing the http endpoint of the Blazor development server it works as expected

I've tried this on both Windows and macOS with the same result

**To Reproduce**

1. Clone the `https://github.com/Apress/beginning-azure-static-web-apps` repo
2. Open the `chapter-5` folder
3. Run `swa start https://localhost:5000 --api-location ./Api --run "dotnet watch run --project ./Client"`
4. Open `http://localhost:4280/`
5. SWA CLI quits without error (in the most recent versions of Node)

**Expected behavior**
The application should load in the browser as it does with the http endpoint

**Desktop (please complete the following information):**

- OS: Windows 11, macOS Ventura 13.0
- SWA CLI Version: 1.0.3

**Additional context**
When downgrading the SWA CLI to 1.0.2 the application works as expected

I tried downgrading Node to see if that had an impact. It did - but only in that running Node 14.x there is an error message displayed. If I build and run the SWA CLI on my machine removing the code that gracefully shuts down the SWA CLI then I can also get this error message on Node 16 (I didn't try on 18.x)

Downgrading Node to 14.21.1 I get the following:

``` ps
(node:12156) UnhandledPromiseRejectionWarning: TypeError [ERR_INVALID_PROTOCOL]: Protocol "https:" not supported. Expected "http:"
[swa]     at new NodeError (internal/errors.js:322:7)
[swa]     at new ClientRequest (_http_client.js:159:11)
[swa]     at Object.request (https.js:370:10)
[swa]     at Array.stream (C:\code\github\static-web-apps-cli\node_modules\http-proxy\lib\http-proxy\passes\web-incoming.js:126:74)
[swa]     at ProxyServer.<anonymous> (C:\code\github\static-web-apps-cli\node_modules\http-proxy\lib\http-proxy\index.js:81:21)
[swa]     at serveStaticOrProxyResponse (C:\code\github\static-web-apps-cli\dist\msha\middlewares\request.middleware.js:97:18)
[swa]     at requestMiddleware (C:\code\github\static-web-apps-cli\dist\msha\middlewares\request.middleware.js:246:9)
[swa]     at Server.<anonymous> (C:\code\github\static-web-apps-cli\dist\msha\server.js:42:58)
[swa]     at Server.emit (events.js:400:28)
[swa]     at parserOnIncoming (_http_server.js:900:12)
[swa]     at HTTPParser.parserOnHeadersComplete (_http_common.js:127:17)
```

On Node 16.18.1 and 18.12.1 I get the following 

``` ps
[swa] remote dev server detected.
[swa]  - url: /
[swa]  - target: https://localhost:5000
[swa] customUrl: false
[swa] is4xx: false
[swa] remote dev server detected. Proxying request
[swa]  - url: /
[swa]  - code: 200
[swa] GET https://localhost:5000/ (proxy)
[swa] killing SWA CLI
[swa] node "C:\Users\s.cashmore\AppData\Roaming\npm\node_modules\@azure\static-web-apps-cli\dist\msha\server.js" exited with code 0
```

**Comment by anthonychu** [comment]
That's due to explicitly using the HTTP agent here, it doesn't work for HTTPS: https://github.com/Azure/static-web-apps-cli/blob/2cd0e980ea4bf604df10fd5d63149b20b292e916/src/msha/server.ts#L20-L23

Not sure if there's a way to provide separate agents for HTTP and HTTPS.

**Comment by StacyCash** [comment]
Link to Twitter discussion of bug

https://twitter.com/Stacy_Cash/status/1595445964751372294?s=20&t=vWEfOcD-NB8VWPucXjT7MA

**Comment by manekinekko** [comment]
I managed to reproduce this issue! 

Actually, there are multiple issues related issues here:

1. As @StacyCash mentioned in this issue, this happens when a user starts the SWA CLI in HTTP, but tries to proxy to a devserver in HTTPS --> In this situation, I'd recommend we simply print a warning asking the user to start the SWA CLI using `--ssl` config and quit. The reason for this is that I'd concerned the user ends up in a situation of [Mixed Active Content (MAC)](https://developer.mozilla.org/en-US/docs/Web/Security/Mixed_content#mixed_active_content) and the browser will eventually block some resources. 
   _Disclaimer: I haven't tested this use case. MAC happens when the user visiting a page over HTTPS tries to load content over HTTP. I'm not 100% sure this is related here and needs to be proved that MAC can happen here._
3. The other issue related to what @anthonychu pointed out is that we are always using `http.Agent()` to proxy requests. This is an issue and we need to make sure `https.Agent()` if the user started SWA CLI using `--ssl`.

**Comment by manekinekko** [comment]
> Aside from the worry of the MAC issue, doi you have anything against the idea of @anthonychu to use the dev server URL to decide whether to use https or not?

I am not against it. If MAC is not an issue here, then @anthonychu's suggestion is part of the fix.

**Comment by StacyCash** [comment]
As a quick test, I've just implemented the solution proposed by @anthonychu and built it locally and it does seem to restore the same functionality that was available in 1.0.2 I can use both http and https dev servers as I could before.

I will try to do some more extensive testing tomorrow (though it will be limited to Blazor WASM)

**Comment by manekinekko** [comment]
Awesome! Could you also try with examples from https://developer.mozilla.org/en-US/docs/Web/Security/Mixed_content#loading_locally_delivered_mixed-resources?

**Comment by StacyCash** [comment]
Sorry, life got in the way on Friday...

I've tested the proposed solution using:

- DotNet calling a DotNet function returning fixed code
- DotNet calling a DotNet function returning external data
- Published DotNet static files calling a DotNet function returning external code
- Angular calling a TypeScript function returning external data

I've tried this on both a MacBook Pro M1 on the latest macOS and Windows 11 22H2

I didn't get any errors, and looking at the console got not MAC errors in the console. It treats the connection as insecure no matter what dev server I'm using

**Comment by eltorio** [comment]
@manekinekko nice to hear that this issue is going to be solved in a near future.
There is many usage where we need https (or bypass security with obscure flags)

**Comment by sgollapudi77** [comment]
We released a new SWA_CLI version(1.0.6) last week with the fix for this, feel free to re-open this if you're still facing this issue.

**Comment by StacyCash** [comment]
Hi, the linked solution doesn't solve this problem. I'm not trying (nor do I want) to run using SSL. I am just trying to connect to a development server that uses HTTPS. The fix suggested by @anthonychu solves the problem correctly. (I made the test to see if the solution worked, but as it wasn't my solution I didn't make a PR for it - I can do if needed)

Can we reopen this issue please?

**Comment by pmarangoni** [comment]
> As a quick test, I've just implemented the solution proposed by [@anthonychu](https://github.com/anthonychu) and built it locally and it does seem to restore the same functionality that was available in 1.0.2 I can use both http and https dev servers as I could before.
> 
> I will try to do some more extensive testing tomorrow (though it will be limited to Blazor WASM)

What is the solution?

**Comment by StacyCash** [comment]
@pmarangoni I haven't looked into the code in a while so I am not sure if this would still be the real fix, but this comment by @anthonychu showed a fix.

https://github.com/Azure/static-web-apps-cli/issues/610#issuecomment-1325596711

The work around is to use the .NET gttp dev server endpoint, rather than the https.

I've I've misunderstood what you are asking, feel free to correct me 😊
